File: backend/goat_listing/src/main/java/com/goat/cart/application/usecases/ClearCartUseCase.java
package com.goat.cart.application.usecases;

import com.goat.cart.adapters.external.RestCartServiceAdapter;
import com.goat.cart.ports.CartRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.UUID;

/**
 * Use case para vaciar el carrito de un usuario.
 * 
 * Usa el adaptador HTTP directamente para comunicarse con el servicio Python,
 * manteniendo la arquitectura hexagonal (el adaptador está en la capa de infraestructura).
 */
@Component
public class ClearCartUseCase {

    private static final Logger logger = LoggerFactory.getLogger(ClearCartUseCase.class);

    private final CartRepository cartRepository;
    private final RestCartServiceAdapter cartServiceAdapter;

    public ClearCartUseCase(CartRepository cartRepository, RestCartServiceAdapter cartServiceAdapter) {
        this.cartRepository = cartRepository;
        this.cartServiceAdapter = cartServiceAdapter;
    }

    public void execute(UUID userId) {
        logger.info("Vaciando carrito del usuario {}", userId);

        // 1. Verificar que el carrito existe
        var cart = cartRepository.findActiveByUser(userId);

        if (cart.isEmpty()) {
            logger.info("El carrito del usuario {} ya estaba vacío", userId);
            return;
        }

        // 2. Vaciar el carrito directamente en el servicio Python
        cartServiceAdapter.clearCart(userId);

        logger.info("Carrito del usuario {} vaciado", userId);
    }
}

File: backend/goat_catalog/src/goat_catalog/cart/application/use_cases/clear_cart_use_case.py
"""Caso de uso para vaciar el carrito."""

import logging
from uuid import UUID

from ...domain.repositories import CartRepository

logger = logging.getLogger(__name__)


class ClearCartUseCase:
    """Caso de uso para vaciar el carrito de un usuario."""

    def __init__(self, cart_repository: CartRepository) -> None:
        """Inicializa el caso de uso.

        Args:
            cart_repository: Repositorio para acceder a los carritos
        """
        self._cart_repository = cart_repository

    async def execute(self, user_id: UUID) -> None:
        """Vacía el carrito del usuario.

        Args:
            user_id: ID del usuario
        """
        logger.info("Vaciando carrito del usuario %s", user_id)

        # 1. Obtener carrito del usuario
        cart = await self._cart_repository.find_active_by_user(user_id)

        if cart is None or cart.is_empty():
            logger.info("El carrito del usuario %s ya estaba vacío", user_id)
            return

        # 2. Limpiar items
        cart.clear()

        # 3. Guardar cambios
        await self._cart_repository.save(cart)

        logger.info("Carrito del usuario %s vaciado", user_id)

